File: app/src/main/java/com/gnest/remember/view/fragments/SettingsFragment.java
package com.gnest.remember.view.fragments;

import android.annotation.TargetApi;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentActivity;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.preference.PreferenceManager;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.gnest.remember.App;
import com.gnest.remember.R;
import com.squareup.leakcanary.RefWatcher;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class SettingsFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener {

    private static final int REQUEST_CODE_NOTIFICATION_SOUND = 1;
    private static final int REQUEST_CODE_CHANNEL_SETUP = 2;

    @BindView(R.id.settings_fragment_toolbar)
    Toolbar toolbar;

    private static SharedPreferences SHARED_PREFERENCES;

    private OnSettingsFragmentInteractionListener mListener;
    private Unbinder mUnbinder;

    public static SettingsFragment newInstance() {
        return new SettingsFragment();
    }

    public SettingsFragment() {
    }

    @Override
    public void onCreatePreferences(@Nullable Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preferences, rootKey);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        RefWatcher refWatcher = App.getRefWatcher();
        refWatcher.watch(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        if (view != null) {
            mUnbinder = ButterKnife.bind(this, view);
        }
        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnSettingsFragmentInteractionListener) {
            mListener = (OnSettingsFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnSettingsFragmentInteractionListener");
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        AppCompatActivity activity = ((AppCompatActivity) getActivity());
        if (activity != null) {
            activity.setSupportActionBar(toolbar);
            ActionBar actionBar = activity.getSupportActionBar();
            if (actionBar != null) {
                actionBar.setDisplayHomeAsUpEnabled(true);
            }
            SHARED_PREFERENCES = PreferenceManager.getDefaultSharedPreferences(activity);
            SHARED_PREFERENCES.registerOnSharedPreferenceChangeListener(this);
        }
        setFontPrefSummary();
        setNotificationSoundPrefSummary(getRingtoneUri());
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnbinder.unbind();
        SHARED_PREFERENCES.unregisterOnSharedPreferenceChangeListener(this);
        SHARED_PREFERENCES = null;
    }

    @Override
    public boolean onPreferenceTreeClick(Preference preference) {
        if (App.NOTIFICATION_SOUND_KEY.equals(preference.getKey())) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                openNotificationChanelSetup();
            } else {
                openRingtonePicker();
            }
            return true;
        } else {
            return super.onPreferenceTreeClick(preference);
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private void openNotificationChanelSetup() {
        Context context = getContext();
        if (context != null) {
            Intent intent = new Intent(Settings.ACTION_CHANNEL_NOTIFICATION_SETTINGS);
            intent.putExtra(Settings.EXTRA_APP_PACKAGE, context.getPackageName());
            intent.putExtra(Settings.EXTRA_CHANNEL_ID, App.NOTIFICATION_CHANNEL_ID);
            startActivityForResult(intent, REQUEST_CODE_CHANNEL_SETUP);
        }
    }

    private void openRingtonePicker() {
        Intent intent = new Intent(RingtoneManager.ACTION_RINGTONE_PICKER);
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, RingtoneManager.TYPE_NOTIFICATION);
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT, true);
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT, true);
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI, Settings.System.DEFAULT_NOTIFICATION_URI);

        String notificationSoundPath = SHARED_PREFERENCES.getString(App.NOTIFICATION_SOUND_KEY, null);
        if (notificationSoundPath != null) {
            if (notificationSoundPath.length() == 0) {
                // Select "Silent"
                intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, "");
            } else {
                intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, Uri.parse(notificationSoundPath));
            }
        } else {
            // No ringtone has been selected, set to the default
            intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, Settings.System.DEFAULT_NOTIFICATION_URI);
        }

        startActivityForResult(intent, REQUEST_CODE_NOTIFICATION_SOUND);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_CODE_NOTIFICATION_SOUND && data != null) {
            Uri ringtone = data.getParcelableExtra(RingtoneManager.EXTRA_RINGTONE_PICKED_URI);
            SharedPreferences.Editor editor = SHARED_PREFERENCES.edit();
            if (ringtone != null) {
                editor.putString(App.NOTIFICATION_SOUND_KEY, ringtone.toString());
            } else {
                // "Silent" was selected
                editor.putString(App.NOTIFICATION_SOUND_KEY, "");
            }
            editor.apply();
        } else if (requestCode == REQUEST_CODE_CHANNEL_SETUP) {
            setNotificationSoundPrefSummary(getRingtoneUri());
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                mListener.onBackPressed();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        switch (key) {
            case App.FONT_SIZE_KEY:
                String fontSize = setFontPrefSummary();
                App.setFontSize(Integer.parseInt(fontSize));
                break;
            case App.NOTIFICATION_SOUND_KEY:
                Uri ringtoneUri = getRingtoneUri();
                setNotificationSoundPrefSummary(ringtoneUri);
                App.setNotificationSound(ringtoneUri);
        }
    }

    @NonNull
    private String setFontPrefSummary() {
        Preference fontSizePref = findPreference(App.FONT_SIZE_KEY);
        String fontSize = SHARED_PREFERENCES.getString(App.FONT_SIZE_KEY, "");
        fontSizePref.setSummary(fontSize);
        return fontSize;
    }

    private void setNotificationSoundPrefSummary(Uri ringtoneUri) {
        Preference notificationSoundPref = findPreference(App.NOTIFICATION_SOUND_KEY);
        Context context = getContext();
        if (context != null) {
            String summary = RingtoneManager.getRingtone(context, ringtoneUri).getTitle(context);
            notificationSoundPref.setSummary(summary);
        }
    }

    private Uri getRingtoneUri() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            FragmentActivity activity = getActivity();
            if (activity != null) {
                NotificationManager manager = (NotificationManager) activity.getSystemService(Context.NOTIFICATION_SERVICE);
                if (manager != null) {
                    NotificationChannel channel = manager.getNotificationChannel(App.NOTIFICATION_CHANNEL_ID);
                    if (channel != null) {
                        return channel.getSound();
                    }
                }
            }
        }
        String uriPath = SHARED_PREFERENCES.getString(App.NOTIFICATION_SOUND_KEY, "");
        return uriPath.isEmpty() ? Settings.System.DEFAULT_NOTIFICATION_URI : Uri.parse(uriPath);
    }


    public interface OnSettingsFragmentInteractionListener {
        void onBackPressed();
    }
}

File: app/src/main/java/com/gnest/remember/ui/fragments/SettingsFragment.kt
package com.gnest.remember.ui.fragments

import android.annotation.TargetApi
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.media.RingtoneManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.*
import androidx.activity.OnBackPressedCallback
import androidx.core.content.edit
import androidx.navigation.fragment.findNavController

import com.gnest.remember.App
import com.gnest.remember.R

import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.gnest.remember.extensions.setSupportActionBar
import com.gnest.remember.extensions.setupActionBarWithNavController
import com.gnest.remember.extensions.sharedPreferences
import com.gnest.remember.extensions.supportActionBar
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.fragment_settings.*

class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {

    private val ringtoneUri: Uri
        get() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val activity = activity
                if (activity != null) {
                    val manager = requireContext().getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager
                    manager?.let {
                        val channel = manager.getNotificationChannel(App.NOTIFICATION_CHANNEL_ID)
                        channel?.let { return it.sound }
                    }
                }
            }
            val uriPath = sharedPreferences().getString(App.NOTIFICATION_SOUND_KEY, "")
            return if (uriPath?.isEmpty() != false) Settings.System.DEFAULT_NOTIFICATION_URI else Uri.parse(uriPath)
        }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
//        val refWatcher = App.getRefWatcher()
//        refWatcher.watch(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setSupportActionBar(settingsFragmentToolbar)
        supportActionBar()?.setDisplayHomeAsUpEnabled(true)
        supportActionBar()?.setHomeButtonEnabled(true)
        activity?.drawerLayout?.let { setupActionBarWithNavController(it) }
        sharedPreferences().registerOnSharedPreferenceChangeListener(this)
//        activity?.addOnBackPressedCallback(this)
        setFontPrefSummary()
        setNotificationSoundPrefSummary(ringtoneUri)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        sharedPreferences().unregisterOnSharedPreferenceChangeListener(this)
//        activity?.removeOnBackPressedCallback(this)
    }

//    override fun handleOnBackPressed(): Boolean {
//        findNavController().navigateUp()
//        return true
//    }

    override fun onPreferenceTreeClick(preference: Preference): Boolean {
        return if (App.NOTIFICATION_SOUND_KEY == preference.key) {
            when {
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O -> openNotificationChanelSetup()
                else -> openRingtonePicker()
            }
            true
        } else {
            super.onPreferenceTreeClick(preference)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when {
            requestCode == REQUEST_CODE_NOTIFICATION_SOUND && data != null -> {
                val ringtone = data.getParcelableExtra<Uri>(RingtoneManager.EXTRA_RINGTONE_PICKED_URI)
                sharedPreferences().edit {
                    ringtone?.let { putString(App.NOTIFICATION_SOUND_KEY, ringtone.toString()) }
                            ?: putString(App.NOTIFICATION_SOUND_KEY, "") // "Silent" was selected
                }
            }
            requestCode == REQUEST_CODE_CHANNEL_SETUP -> setNotificationSoundPrefSummary(ringtoneUri)
            else -> super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        android.R.id.home -> {
            findNavController().navigateUp()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        when (key) {
            App.FONT_SIZE_KEY -> {
                val fontSize = setFontPrefSummary()
                App.setFontSize(Integer.parseInt(fontSize))
            }
            App.NOTIFICATION_SOUND_KEY -> {
                val ringtoneUri = ringtoneUri
                setNotificationSoundPrefSummary(ringtoneUri)
                App.setNotificationSound(ringtoneUri)
            }
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun openNotificationChanelSetup() {
        val intent = Intent(Settings.ACTION_CHANNEL_NOTIFICATION_SETTINGS)
        intent.putExtra(Settings.EXTRA_APP_PACKAGE, requireContext().packageName)
        intent.putExtra(Settings.EXTRA_CHANNEL_ID, App.NOTIFICATION_CHANNEL_ID)
        startActivityForResult(intent, REQUEST_CODE_CHANNEL_SETUP)
    }

    private fun openRingtonePicker() {
        val intent = Intent(RingtoneManager.ACTION_RINGTONE_PICKER)
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, RingtoneManager.TYPE_NOTIFICATION)
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_DEFAULT, true)
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_SHOW_SILENT, true)
        intent.putExtra(RingtoneManager.EXTRA_RINGTONE_DEFAULT_URI, Settings.System.DEFAULT_NOTIFICATION_URI)

        val notificationSoundPath = sharedPreferences().getString(App.NOTIFICATION_SOUND_KEY, null)

        notificationSoundPath?.let {
            when {
                it.isEmpty() -> // Select "Silent"
                    intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, "")
                else -> intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, Uri.parse(notificationSoundPath))
            }
        } ?: run {
            // No ringtone has been selected, set to the default
            intent.putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, Settings.System.DEFAULT_NOTIFICATION_URI)
        }

        startActivityForResult(intent, REQUEST_CODE_NOTIFICATION_SOUND)
    }

    private fun setFontPrefSummary(): String {
        val fontSizePref = findPreference(App.FONT_SIZE_KEY)
        val fontSize = sharedPreferences().getString(App.FONT_SIZE_KEY, "") ?: ""
        fontSizePref.summary = fontSize
        return fontSize
    }

    private fun setNotificationSoundPrefSummary(ringtoneUri: Uri) {
        val notificationSoundPref = findPreference(App.NOTIFICATION_SOUND_KEY)
        val summary = RingtoneManager.getRingtone(requireContext(), ringtoneUri).getTitle(context)
        notificationSoundPref.summary = summary
    }

    companion object {

        private const val REQUEST_CODE_NOTIFICATION_SOUND = 1
        private const val REQUEST_CODE_CHANNEL_SETUP = 2
    }
}